tests/l2_rpc.rs

```diff
@@ -0,0 +1,114 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+pub mod test_lib;
+
+use std::{str::FromStr, sync::Arc};
+use tokio::time::{sleep, Duration};
+use up_rust::{
+    communication::{
+        CallOptions, InMemoryRpcServer, RequestHandler, RpcClient, RpcServer,
+        ServiceInvocationError, UPayload,
+    },
+    LocalUriProvider, UPayloadFormat, UUri,
+};
+use up_transport_zenoh::ZenohRpcClient;
+
+const METHOD_RESOURCE_ID: u16 = 0x00a0;
+
+struct ExampleHandler {
+    request_data: String,
+    response_data: String,
+}
+impl ExampleHandler {
+    pub fn new(request_data: String, response_data: String) -> Self {
+        ExampleHandler {
+            request_data,
+            response_data,
+        }
+    }
+}
+#[async_trait::async_trait]
+impl RequestHandler for ExampleHandler {
+    async fn invoke_method(
+        &self,
+        _resource_id: u16,
+        request_payload: Option<UPayload>,
+    ) -> Result<Option<UPayload>, ServiceInvocationError> {
+        let payload = request_payload.unwrap().payload();
+        let data = payload.into_iter().map(|c| c as char).collect::<String>();
+        // compare the request data
+        assert_eq!(data, self.request_data);
+        // return
+        let payload = UPayload::new(
+            self.response_data.clone().into(),
+            UPayloadFormat::UPAYLOAD_FORMAT_TEXT,
+        );
+        Ok(Some(payload))
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_l2_rpc() {
+    test_lib::before_test();
+
+    // Initialization
+    let request_data = String::from("This is the request data");
+    let response_data = String::from("This is the response data");
+    let uptransport_server = Arc::new(
+        test_lib::create_up_transport_zenoh("//l2_responder/2/1/0")
+            .await
+            .unwrap(),
+    );
+    let uptransport_client = Arc::new(
+        test_lib::create_up_transport_zenoh("//l2_requester/1/1/0")
+            .await
+            .unwrap(),
+    );
+
+    // Create L2 RPC server
+    let rpc_server = InMemoryRpcServer::new(uptransport_server.clone(), uptransport_server.clone());
+    let example_handler = Arc::new(ExampleHandler::new(
+        request_data.clone(),
+        response_data.clone(),
+    ));
+    // CY_TODO: Verify the what kind of resource_id we can listen to
+    rpc_server
+        .register_endpoint(
+            Some(&UUri::from_str("//*/FFFF/FF/0").unwrap()),
+            METHOD_RESOURCE_ID,
+            example_handler.clone(),
+        )
+        .await
+        .unwrap();
+    // Need some time for queryable to run
```

Comment:
That means Zenoh queryable, which is used in RPC server.

Reply:
My understanding is that UPZenohTransport declares a Zenoh Queryable for the endpoint registered by InMemoryRpcServer, right? And it might take a while to _announce_ the queryable across the Zenoh network? So to be sure that a request can be handled, the client needs to wait a little?

Reply:
Yes, you're right.

Reply:
Is there a way for the transport to know when the queryable has been announced on the network, e.g. by means of a callback or something like that? Waiting for an undefined amount of time does not look very deterministic to me ...

Reply:
Hmm.. we don't have such a way to know this right now. But I think 1 second is pretty enough in this case. Let me loop @Mallets if you want to add any comments.